### ISSUE TITLE
SC2181 conflicts with SC2310

### ISSUE THREAD

**Issue by nihil-admirari** [koalaman/shellcheck#2709]
#### For bugs
- Rule Ids: SC2181 and SC2310
- My shellcheck version: 0.9.0
- [x] The rule's wiki page does not already cover this
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
#shellcheck enable=check-set-e-suppressed

if OUT=$(
    set -euo pipefail

    success() {
        echo success;
    }

    fail() {
        echo fail
        return 1
    }

    out=$(fail)
    out+=$(success)
    out+=$(fail)

    printf '%s' "$out"
); then
    export OUT
fi
```

#### Here's what shellcheck currently says:

> Line 16:
>     out=$(fail)
>           ^-- [SC2310](https://www.shellcheck.net/wiki/SC2310) (info): This function is invoked in an 'if' condition so set -e will be disabled. Invoke separately if failures should cause the script to exit.

### Rewriting as suggested:
```sh
#!/bin/bash
#shellcheck enable=check-set-e-suppressed

OUT=$(
    set -euo pipefail

    success() {
        echo success;
    }

    fail() {
        echo fail
        return 1
    }

    out=$(fail)
    out+=$(success)
    out+=$(fail)

    printf '%s' "$out"
)

if (( $? == 0 )); then
    export OUT
fi
```

gives  SC2181:

> Line 23:
> if (( $? == 0 )); then
>       ^-- [SC2181](https://www.shellcheck.net/wiki/SC2181) (style): Check exit code directly with e.g. 'if mycmd;', not indirectly with $?.

#### Here's what I wanted or expected to see:

SC2181 is not issued if the last command is prone to SC2310 (or any other warning).

**Comment by h3xx** [comment]
Because you have `set -e` on, the `if` statement becomes unnecessary. The script dies after it finishes executing line 15, so I think the suggestion was actually to re-write the script as:

```sh
#!/bin/bash
#shellcheck enable=check-set-e-suppressed

set -euo pipefail

success() {
    echo success;
}

fail() {
    echo fail
    return 1
}

out=$(fail; success; fail)
echo "$out"
```

**Comment by nihil-admirari** [comment]
Sorry, an example from a bug report is a mistranslation of an actual script that gave such an errot. In the original script, `set -e` was active only in a subshell:

```bash
OUT=$(
    set -euo pipefail
   ...
)
```

I've updated the example. Please let me now if it still rewritable in a more idiomatic way, which doesn't give a SC2181/SC2310 conflict.

**Comment by nihil-admirari** [comment]
Ok, I'll close this one as a duplicate of #2738 to reduce the number of open issues.
